packages/vm/src/cache.rs

```diff
@@ -161,42 +181,110 @@ mod test {
+        }
+    }
+
+    #[test]
+    fn load_wasm_works() {
+        let tmp_dir = TempDir::new().unwrap();
+        let mut cache: CosmCache<MockStorage, MockApi, MockQuerier> =
+            unsafe { CosmCache::new(tmp_dir.path(), default_features(), 10).unwrap() };
+        let id = cache.save_wasm(CONTRACT).unwrap();
+
+        let restored = cache.load_wasm(&id).unwrap();
+        assert_eq!(restored, CONTRACT);
+    }
+
+    #[test]
+    fn load_wasm_works_across_multiple_instances() {
+        let tmp_dir = TempDir::new().unwrap();
+        let tmp_path = tmp_dir.path();
+        let id: Checksum;
+
+        {
+            let mut cache1: CosmCache<MockStorage, MockApi, MockQuerier> =
+                unsafe { CosmCache::new(tmp_path, default_features(), 10).unwrap() };
+            id = cache1.save_wasm(CONTRACT).unwrap();
+        }
 
-        // this is invalid, as it doesn't contain all required exports
-        static WAT: &'static str = r#"
-            (module
-              (type $t0 (func (param i32) (result i32)))
-              (func $add_one (export "add_one") (type $t0) (param $p0 i32) (result i32)
-                get_local $p0
-                i32.const 1
-                i32.add))
-        "#;
+        {
+            let cache2: CosmCache<MockStorage, MockApi, MockQuerier> =
+                unsafe { CosmCache::new(tmp_path, default_features(), 10).unwrap() };
+            let restored = cache2.load_wasm(&id).unwrap();
+            assert_eq!(restored, CONTRACT);
+        }
+    }
 
-        let wasm = wat2wasm(WAT).unwrap();
+    #[test]
+    fn load_wasm_errors_for_non_existent_id() {
+        let tmp_dir = TempDir::new().unwrap();
+        let cache: CosmCache<MockStorage, MockApi, MockQuerier> =
+            unsafe { CosmCache::new(tmp_dir.path(), default_features(), 10).unwrap() };
+        let checksum = Checksum::from([
+            5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
+            5, 5, 5,
+        ]);
+
+        let res = cache.load_wasm(&checksum);
+        match res {
+            Err(VmError::IoErr { .. }) => {}
+            Err(e) => panic!("Unexpected error: {:?}", e),
+            Ok(_) => panic!("This must not succeed"),
+        }
+    }
 
+    #[test]
+    fn load_wasm_errors_for_corrupted_wasm() {
         let tmp_dir = TempDir::new().unwrap();
         let mut cache: CosmCache<MockStorage, MockApi, MockQuerier> =
-            unsafe { CosmCache::new(tmp_dir.path(), 10).unwrap() };
-        let save_result = cache.save_wasm(&wasm);
-        match save_result {
-            Err(VmError::ValidationErr { .. }) => {}
-            Err(e) => panic!("Unexpected error {:?}", e),
-            Ok(_) => panic!("Didn't reject wasm with invalid api"),
+            unsafe { CosmCache::new(tmp_dir.path(), default_features(), 10).unwrap() };
+        let checksum = cache.save_wasm(CONTRACT).unwrap();
+
+        // Corrupt cache file
```

Comment:
Ah, this is now validated at load time. Or at least that is tested.
Good enhancement